# <h1><b>Battle of Neighborhoods - A Coursera Capstone project (Week 2) </b></h1>

<!-- ## Table of contents
* [Introduction](#introduction)
* [Data](#data)
* [Data_Sources](#data sources)
* [Methodology](#methodology)
* [Analysis](#analysis)
* [Results and Discussion](#results)
* [Conclusion](#conclusion) -->

## **Introduction** <a name="introduction"></a>

Arizona and Hawaii- both the states in the United States of America have "state of the art" telescopes and observatories. As a grad student who wants to pursue a PhD in Astronomy, these giant visible-light range telescopes are subjects of fascination. For this capstone project, I'll explore the neighborhoods of Hawaii as well as Arizona and cluster the neighborhoods to show similarities and dissimilarities between these two places. 

Hopefully, **students, faculty members and job-seeking candidates who want to work at the observatories of either of the mentioned places will be benefitted from this project as they will be able to make decisions about whether to stay near this place or travel from a further location, based on their preferences.**

## <b>Data</b><a name="data"></a>

### Description of the data that will be used in this project-

 - **List of neighborhoods in Hawaii with their latitudes and longitudes**. The types of venues have also been extracted so that segmentation and clustering is easier. 
 - **List of neighborhoods in Arizona with their latitudes and longitudes** and similarly the types of venues have also been extracted.
 - **Location data from the Foursquare API to segment and cluster the neighborhoods.**

## **Data Sources** <a name="data sources"></a>

In this project, **Keck Observatory** in **Hawaii** and **Steward Observatory** at the University of Arizona in **Arizona** will be used as the centers, around which the neighborhoods will be explored.

- **Geodata for Hawaii,extracted from a GeoJson file from NYU Spatial Data Repository by loading the .json file.**
- **Geodata for Arizona, extracted from a GeoJson file from NYU Spatial Data Repository, extracted from a GeoJson file from NYU Spatial Data Repository by loading the .json file.**

While downloading the Geojson file, it is important to check if the file contains "point"/"multipoint"
 features. Otherwise, in case of "polygon" features, extracting the types and names of neighborhoods and types of venues becomes much complicated.
 - **Geocode information from Geopy.** 
 - **Location data from the Foursquare API to segment and cluster the neighborhoods. For this segment, the CLIENT_ID and the CLIENT_SECRET is required. It is also required to specify the RADIUS  upto which distance around the specified location the "exploring neighborhood" process will take place.**

Now we'll be gathering the required data to  explore the different neighborhoods around the mentioned centers.

## **Methodology**

**As the 1st step**, we'll need to import different libraries for the analysis. Then we will load the .json or the GeoJson file for Hawaii and import a dataframe with all the required Features,i.e. "Neighborhood","Latitude" and "Longitude", from it. After cleaning the dataframe, we create a Map of Hawaii with Folium and superimpose the neighborhoods on it. 

We then do the same thing for Arizona's Steward Observatory. We load the Geojson file and then import a dataframe from it with all the required Features,i.e. "Neighborhood","Latitude" and "Longitude". Then we create a map of Arizona with Folium and superimpose the markers on it.

**In the 2nd step**, we explore the neighborhoods around the mentioned centers in both Hawaii and Arizona using the *Foursqaure API/venues/explore*. We will be needing our Foursquare CLIENT_ID and CLIENT_SECRET. We will get another .json file from the **requests** library and will explore the neighborhoods for venues and will get the type of every single venue by defining **get_category_type**. After that, we will get the most common venues around the particular center by defining a specific function for that.

**In the 3rd step**, we will segment and cluster the neighborhoods around both the centers in Hawaii and Arizona. The clusters will be shown on the map of Hawaii and Arizona. For this step, we will use **K-means clustering** from *SciKit-learn* package.


**After completing the clustering method, we will analyze the neighborhoods of each of the places and then suggest a solution of the mentioned problem in the introduction.**



#### **Importing required libraries-**

In [87]:
import  numpy as np

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import folium # map rendering library

## **Analysis**

As we need to explore the neighborhoods around Hawaii's Mauna Kea Observatory and Arizona's Steward Observatory, the Geodata of these places via a Geojson file is required. These Geojson files will be imported as **Hawaii.json** for Hawaii and **Arizona.json** for Arizona.

#### **Exploring Hawaii's data-**

First, we will be exploring the neighborhoods around Hawaii's Keck observatory.

To do that, we need to load the **Hawaii.json** file and import it as **hawaii_data**.

In [88]:
with open('Hawaii.json') as json_data:
    hawaii_data = json.load(json_data)

Now, we'll take a look at **hawaii_data**.

In [5]:
hawaii_data

{'type': 'FeatureCollection',
 'totalFeatures': 936,
 'features': [{'type': 'Feature',
   'id': 'TG00HILPT.1',
   'geometry': {'type': 'MultiPoint',
    'coordinates': [[-156.036222, 19.782857]]},
   'geometry_name': 'the_geom',
   'properties': {'GIST_ID': 1, 'CFCC': 'D82', 'NAME': ''}},
  {'type': 'Feature',
   'id': 'TG00HILPT.2',
   'geometry': {'type': 'MultiPoint',
    'coordinates': [[-156.003466, 19.64056]]},
   'geometry_name': 'the_geom',
   'properties': {'GIST_ID': 2,
    'CFCC': 'D71',
    'NAME': 'Kukailimoku Point Lighthouse'}},
  {'type': 'Feature',
   'id': 'TG00HILPT.3',
   'geometry': {'type': 'MultiPoint',
    'coordinates': [[-156.011014, 19.64701401]]},
   'geometry_name': 'the_geom',
   'properties': {'GIST_ID': 3, 'CFCC': 'D51', 'NAME': 'Old Kona Airport'}},
  {'type': 'Feature',
   'id': 'TG00HILPT.4',
   'geometry': {'type': 'MultiPoint',
    'coordinates': [[-156.043785, 19.73810299]]},
   'geometry_name': 'the_geom',
   'properties': {'GIST_ID': 4,
    'CFCC

As we can see after loading the data that all the relevant data is in the *features* key, which is basically a list of the neighborhoods. So, we define a new variable that includes this data.

In [6]:
neighborhoods_data = hawaii_data['features']

As we have the data of neighborhoods stored as *neighborhoods_data*, we will have a look at the data.

In [8]:
neighborhoods_data

[{'type': 'Feature',
  'id': 'TG00HILPT.1',
  'geometry': {'type': 'MultiPoint',
   'coordinates': [[-156.036222, 19.782857]]},
  'geometry_name': 'the_geom',
  'properties': {'GIST_ID': 1, 'CFCC': 'D82', 'NAME': ''}},
 {'type': 'Feature',
  'id': 'TG00HILPT.2',
  'geometry': {'type': 'MultiPoint', 'coordinates': [[-156.003466, 19.64056]]},
  'geometry_name': 'the_geom',
  'properties': {'GIST_ID': 2,
   'CFCC': 'D71',
   'NAME': 'Kukailimoku Point Lighthouse'}},
 {'type': 'Feature',
  'id': 'TG00HILPT.3',
  'geometry': {'type': 'MultiPoint',
   'coordinates': [[-156.011014, 19.64701401]]},
  'geometry_name': 'the_geom',
  'properties': {'GIST_ID': 3, 'CFCC': 'D51', 'NAME': 'Old Kona Airport'}},
 {'type': 'Feature',
  'id': 'TG00HILPT.4',
  'geometry': {'type': 'MultiPoint',
   'coordinates': [[-156.043785, 19.73810299]]},
  'geometry_name': 'the_geom',
  'properties': {'GIST_ID': 4,
   'CFCC': 'D85',
   'NAME': 'Ellison S Onizuka Space Center'}},
 {'type': 'Feature',
  'id': 'TG00HILP

To inspect tha data, first we need to know how the data looks like and what its features are.

In [89]:
neighborhoods_data[1]

{'type': 'Feature',
 'id': 'TG00HILPT.2',
 'geometry': {'type': 'MultiPoint', 'coordinates': [[-156.003466, 19.64056]]},
 'geometry_name': 'the_geom',
 'properties': {'GIST_ID': 2,
  'CFCC': 'D71',
  'NAME': 'Kukailimoku Point Lighthouse'}}

So, we see that the data contains the name od the neghborhood and its coordinates(Latitude and Longitude). 

We need to convert the features data into a dataframe so that we can do exploratory analysis on it.

We have all the essential features such as **"Neighborhood", "Latitude" and "Longitude"**. So we create an empty dataframe and then include all the features as column names and then include the data from **neighborhoods_data**.

In [90]:
column_names = ['Neighborhood', 'Latitude', 'Longitude'] 

neighborhoods = pd.DataFrame(columns=column_names) # An empty dataframe created.
neighborhoods

Empty DataFrame
Columns: [Neighborhood, Latitude, Longitude]
Index: []

In [91]:
for data in neighborhoods_data:
    neighborhood_name = data['properties']['NAME']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[0][1]
    neighborhood_lon = neighborhood_latlon[0][0]
    
    neighborhoods = neighborhoods.append({'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)

Now we have filled the dataframe with the required features. 

A look at the dataframe-

In [92]:
neighborhoods.head(20)

Neighborhood   Latitude   Longitude
0                                   19.782857 -156.036222
1     Kukailimoku Point Lighthouse  19.640560 -156.003466
2                 Old Kona Airport  19.647014 -156.011014
3   Ellison S Onizuka Space Center  19.738103 -156.043785
4         Keahole Point Lighthouse  19.730984 -156.063283
5      Honokohau Small Boat Harbor  19.673002 -156.025103
6                   Kailua Airport  19.647014 -156.011014
7                                   19.429102 -154.883290
8                                   19.406082 -154.918390
9                                   19.497396 -154.945666
10                                  19.496888 -154.946080
11                                  19.494312 -154.944567
12                                  19.495428 -154.946177
13                                  19.455674 -154.851338
14                                  19.354163 -154.980159
15                                  19.353694 -154.978452
16                                  19.369826 -154.963457
17                                  19.358588 -154.985719
18                                  19.356824 -154.985697
19                                  19.358305 -154.989167

The dataframe looks correct, but we can see a lot of rows are empty and have just coordinates. We'll have to clean up the dataframe.

Let's see the shape of the dataframe.

In [93]:
s = neighborhoods.shape
s

(936, 3)

In [ ]:
for j in range(0,s[0]):
    if neighborhoods.iloc[j,0]=="":
            neighborhoods.drop(labels = j,axis = 0,inplace = True)
            neighborhoods.reset_index(drop = True,inplace = True)

In [109]:
neighborhoods

Neighborhood   Latitude   Longitude
0      Kukailimoku Point Lighthouse  19.640560 -156.003466
1                  Old Kona Airport  19.647014 -156.011014
2    Ellison S Onizuka Space Center  19.738103 -156.043785
3          Keahole Point Lighthouse  19.730984 -156.063283
4       Honokohau Small Boat Harbor  19.673002 -156.025103
5                    Kailua Airport  19.647014 -156.011014
6                       Water Tower  19.676486 -156.006658
7               Kealakekua Bay Park  19.478702 -155.921985
8                 Kalahiki Cemetery  19.378003 -155.878344
9                    Hookena School  19.390148 -155.881706
10                  Honaunau School  19.453100 -155.881097
11                      Alae School  19.332269 -155.876293
12   Seamountain Ninole Golf Course  19.137274 -155.511145
13           Kaalaiki Landing Strip  19.125878 -155.579529
14       Upper Paauau Landing Strip  19.231650 -155.508958
15           Kaalaiki Landing Strip  19.124004 -155.581519
16                     Naalehu Park  19.063949 -155.585591
17                    Waiohinu Park  19.067627 -155.608703
18    Whittington Beach County Park  19.088420 -155.553528
19                    Mookini Heiau  20.261218 -155.880068
20    Kahola Historical Sites State  20.261218 -155.880068
21                    Upolu Airport  20.269011 -155.860285
22         Kauhola Point Lighthouse  20.249070 -155.773223
23                  Kohala Hospital  20.235929 -155.798770
24                   Halaula School  20.230600 -155.781744
25                 Wainaia Cemetery  20.232458 -155.790555
26                    Waimea School  20.022861 -155.672277
27                    Parker School  20.025564 -155.672325
28               Laupahoehoe School  19.982727 -155.230532
29                   Hakalau School  19.891776 -155.127845
30                  Waikaumalo Park  19.934818 -155.165765
31        Mauna Kea Science Reserve  19.824991 -155.471990
32                Pepeekeo Airstrip  19.848971 -155.115091
33                     Andrade Camp  19.841755 -155.109342
34            Waiakea Middle School  19.699361 -155.079949
35                     Wainaku Camp  19.746773 -155.111575
36                       Lehia Park  19.737560 -155.010354
37               Kealoha Beach Park  19.736654 -155.029962
38                      Post Office  19.621999 -155.951204
39       Holualoa Elementary School  19.618879 -155.951571
40                 Holualoa Library  19.621114 -155.951407
41            Konawaena High School  19.513526 -155.917550
42                    Kona Hospital  19.522743 -155.921403
43                 Higashihara Park  19.554625 -155.939652
44          Mauna Lani Point Resort  19.942333 -155.878681
45            Mauna Loa Observatory  19.539166 -155.579313
46     Kulani Correctional Facility  19.565735 -155.323895
47            Mauna Loa Boys School  19.585732 -155.339780
48                     Keaau School  19.624588 -155.040633
49                 Main Post Office  19.719526 -155.052392
50                      Dolphin Bay  19.731972 -155.093281
51              Maui's Canoe (Rock)  19.730174 -155.092140
52          Federal Office Building  19.728780 -155.091684
53                          Library  19.727733 -155.093594
54                 Hilo High School  19.725542 -155.097402
55     East Hawai'i Cultural Center  19.727599 -155.090801
56          Hawai'i Visitors Bureau  19.726607 -155.089603
57              Haili Church (1824)  19.725051 -155.092540
58   Lyman Museum and Mission House  19.725239 -155.093822
59             Puainako Town Center  19.695160 -155.070059
60              Waiakea High School  19.703064 -155.080130
61   Hawai'i State Department of Ag  19.709193 -155.076265
62   Prince Kuhio Plaza Shop Center  19.700135 -155.067009
63            Hawai'i County Police  19.718593 -155.088193
64                     Kaiko'o Mall  19.720969 -155.081735
65            State Office Building  19.721376 -155.080220
66           County Office Building  19.722191 -155.080941
67       Vi

In [112]:
s1 = neighborhoods.shape

In [116]:
print('The dataframe has {} neighborhoods.'.format(s1[0]
    )
)

The dataframe has 515 neighborhoods.


As we can see after cleaning the dataframe, there are 515 neighborhoods.

Now we create a map of Hawaii with the help of **Folium.** The *Nominatim* will use "hawaii_explorer" as *user_agent*.

In [117]:
address = 'Hawaii, USA'

geolocator = Nominatim(user_agent="hawaii_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Hawaii are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Hawaii are 21.2160437, -157.975203.


now we superimpose the neighborhoods on the top of the map of the Hawaii we created.

In [120]:
map_hawaii= folium.Map(location=[latitude, longitude], zoom_start=12)

# add markers to map
for lat, lng, label in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='green',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_hawaii)  
    
map_hawaii

We'll be exploring around the Keck Observatory, so let's get its coordinates from thte **neighborhoods** dataframe.

In [126]:
neighborhoods[neighborhoods["Neighborhood"]=="Keck Observatory"]

Neighborhood   Latitude   Longitude
239  Keck Observatory  19.829536 -155.474938

Getting the coordinates of the **Keck observatory**-

In [129]:
Latitude = neighborhoods[neighborhoods["Neighborhood"]=="Keck Observatory"].iloc[0,1]
Longitude = neighborhoods[neighborhoods["Neighborhood"]=="Keck Observatory"].iloc[0,2]

In [131]:
neighborhood_name = "Keck Observatory" # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               Latitude, 
                                                               Longitude))

Latitude and longitude values of Keck Observatory are 19.829536, -155.474938.


We will explore around this mentioned neighborhood using thte Foursquare API.

For doing this, we will be needing **CLIENT_ID** and **CLIENT_SECRET** from the Foursquare developer console.

In [170]:
CLIENT_ID = 'HXVAUOWVYH4BQ3FBLW4S2FXKHH0UMJISPKP0VRM0D3KPFSRB' # Foursquare ID
CLIENT_SECRET = "2P4EN1UVGEK5OBMCOYRUFI0CQ1MKP2A1HKPAQ0KASSA2D4XW" #  Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Credentails:
CLIENT_ID: HXVAUOWVYH4BQ3FBLW4S2FXKHH0UMJISPKP0VRM0D3KPFSRB
CLIENT_SECRET:2P4EN1UVGEK5OBMCOYRUFI0CQ1MKP2A1HKPAQ0KASSA2D4XW


For the next step, **venues/explore** on the Foursquare API will be used to explore around the neighborhood.

In [175]:
LIMIT = 100
radius = 7000 #The radius around the center for which distance the neighborhood will be explored.
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    Latitude,
    Longitude,
    radius, 
    LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?&client_id=HXVAUOWVYH4BQ3FBLW4S2FXKHH0UMJISPKP0VRM0D3KPFSRB&client_secret=2P4EN1UVGEK5OBMCOYRUFI0CQ1MKP2A1HKPAQ0KASSA2D4XW&v=20180605&ll=19.829536,-155.474938&radius=7000&limit=100'

It is important to note that we had to use an incredibly large RADIUS to get popular locations around the Keck observatory,because generally the observatories are located at remote places. This helps getting more light from the astronomical bodies and less light polution which interfere with the observations. Also, the observatory is situated on a mountain top so the atmosphere is thin and light scattering by the atmosphere is less.

From the **requests** library, we'll get the results of the names of the neighborhoods around the center.

In [176]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5e56e53995feaf001bacabf9'},
 'response': {'headerLocation': 'Current map view',
  'headerFullLocation': 'Current map view',
  'headerLocationGranularity': 'unknown',
  'totalResults': 7,
  'suggestedBounds': {'ne': {'lat': 19.892536063000065,
    'lng': -155.408092000248},
   'sw': {'lat': 19.766535936999936, 'lng': -155.54178399975203}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4b89ecebf964a520ab5632e3',
       'name': 'Mauna Kea Observatory Complex',
       'location': {'address': 'John A. Burns Way',
        'lat': 19.822871495182433,
        'lng': -155.46963214874268,
        'labeledLatLngs': [{'label': 'display',
          'lat': 19.822871495182433,
          'lng': -155.46963214874268}],
        'distance': 926,
       

In [177]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [179]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues

/home/debaparna/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  This is separate from the ipykernel package so we can avoid doing imports until


name            categories        lat  \
0            Mauna Kea Observatory Complex              Mountain  19.822871   
1                             Coffee Break        Breakfast Spot  19.821434   
2                             Waiau (lake)                  Lake  19.811426   
3                  D's Home Repair Service          Home Service  19.784796   
4                     Rover's Fitness Club  Gym / Fitness Center  19.867388   
5                 Monstera Noodles | Sushi      Asian Restaurant  19.815004   
6  Mauna Kea Ice Age Natural Area Preserve       Nature Preserve  19.769506   

          lng  
0 -155.469632  
1 -155.484480  
2 -155.477700  
3 -155.493622  
4 -155.425321  
5 -155.537851  
6 -155.456116

In [180]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

7 venues were returned by Foursquare.


Now, we will be defining a function which will extract nearby venues from the Foursqaure API.

In [181]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

We will be getting nearby venues of the Keck Observatory now.

In [182]:
hawaii_venues = getNearbyVenues(names=neighborhoods['Neighborhood'],
                                   latitudes=neighborhoods['Latitude'],
                                   longitudes=neighborhoods['Longitude']
                                  )

Kukailimoku Point Lighthouse
Old Kona Airport
Ellison S Onizuka Space Center
Keahole Point Lighthouse
Honokohau Small Boat Harbor
Kailua Airport
Water Tower
Kealakekua Bay Park
Kalahiki Cemetery
Hookena School
Honaunau School
Alae School
Seamountain Ninole Golf Course
Kaalaiki Landing Strip
Upper Paauau Landing Strip
Kaalaiki Landing Strip
Naalehu Park
Waiohinu Park
Whittington Beach County Park
Mookini Heiau
Kahola Historical Sites State
Upolu Airport
Kauhola Point Lighthouse
Kohala Hospital
Halaula School
Wainaia Cemetery
Waimea School
Parker School
Laupahoehoe School
Hakalau School
Waikaumalo Park
Mauna Kea Science Reserve
Pepeekeo Airstrip
Andrade Camp
Waiakea Middle School
Wainaku Camp
Lehia Park
Kealoha Beach Park
Post Office
Holualoa Elementary School
Holualoa Library
Konawaena High School
Kona Hospital
Higashihara Park
Mauna Lani Point Resort
Mauna Loa Observatory
Kulani Correctional Facility
Mauna Loa Boys School
Keaau School
Main Post Office
Dolphin Bay
Maui's Canoe (Rock)
Fe

KeyError: 'groups'

In [183]:
print(hawaii_venues.shape)

(5705, 7)


now let's take look at the venues dataframe.

In [184]:
hawaii_venues.head()

Neighborhood  Neighborhood Latitude  \
0  Kukailimoku Point Lighthouse               19.64056   
1  Kukailimoku Point Lighthouse               19.64056   
2  Kukailimoku Point Lighthouse               19.64056   
3  Kukailimoku Point Lighthouse               19.64056   
4  Kukailimoku Point Lighthouse               19.64056   

   Neighborhood Longitude                              Venue  Venue Latitude  \
0             -156.003466  Umeke's Fish Market, Bar, & Grill       19.642319   
1             -156.003466                Orchid Thai Cuisine       19.642815   
2             -156.003466                          El Maguey       19.642389   
3             -156.003466                        Keiki Beach       19.638980   
4             -156.003466                        Irie Hawaii       19.641475   

   Venue Longitude       Venue Category  
0      -156.000651  Hawaiian Restaurant  
1      -156.000835      Thai Restaurant  
2      -156.000248   Mexican Restaurant  
3      -156.002540                Beach  
4      -156.000211           Smoke Shop

In [185]:
hawaii_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  \
Neighborhood                                                                    
(Pvt) Park                                         20                      20   
11                                                  1                       1   
12                                                  1                       1   
13                                                  2                       2   
A&B Sugar Museum                                   15                      15   
Afook-Chinen Civic Auditorium                      10                      10   
Ahu'ena Heiau                                      27                      27   
Ahukini Rec Pier State Park                         2                       2   
Aiea Heights Rest Home                              2                       2   
Aiea High School                                    1                       1   
Aiea Recreation Center                              1                       1   
Aikahi Elementary                                   2                       2   
Aikahi Playground                                   2                       2   
Aikane Park                                        17                      17   
Akebono Theather                                   12                      12   
Ali'i Sunset Plaza                                 37                      37   
Aloha Tower                                        23                      23   
Alvah A Scott Elementary                           64                      64   
Army Museum                                        81                      81   
Baldwin House                                      47                      47   
Bay View Golf Course                               18                      18   
Benjamin Parker School                             39                      39   
Big Island Country Club                             1                       1   
Burns Field Airstrip                                4                       4   
Bus Station                                        42                      42   
Camp Maluhia                                        1                       1   
Camp Sloggett Ywca                                  1                       1   
Cemetary                                           59                      59   
Civic Center                                       12                      12   
Civic Center Auditorium                            10                      10   
Cmdr & Chief Pacific Fleet                         14                      14   
Convention Hall                                    16                      16   
County Office Building                             23                      23   
Courthouse                                         58                      58   
Creative Day School                                 3                       3   
Crosroads Center                                   11                      11   
De Corte Playground                                 5                       5   
Department of Commerce N I S T                      2                       2   
Diamond Head Civil Defense Hdq                      5                       5   
Dolphin Bay                                         1                       1   
East Hawai'i Cultural Center                       22                      22   
Edith Kankaole Multipurpose St                     15                      15   
Ellison S Onizuka Astronomy Co                      1                       1   
Ellison S Onizuka Space Center                      9                       9   
Enchanted Lake Shopping Center                     19                      19   
Federal Office Building                             5                       5   
Fire Department                                    46                      46   
Fire Station                                       32          

In [186]:
print('There are {} uniques categories.'.format(len(hawaii_venues['Venue Category'].unique())))

There are 282 uniques categories.


So, if we consider a really vast search area from the observatory, it's a very diverse neighborhood and there are a lot of things to do.

In [187]:
# one hot encoding
hawaii_onehot = pd.get_dummies(hawaii_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
hawaii_onehot['Neighborhood'] = hawaii_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [hawaii_onehot.columns[-1]] + list(hawaii_onehot.columns[:-1])
hawaii_onehot = hawaii_onehot[fixed_columns]

hawaii_onehot.head(20)

Neighborhood  Accessories Store  Airport  \
0     Kukailimoku Point Lighthouse                  0        0   
1     Kukailimoku Point Lighthouse                  0        0   
2     Kukailimoku Point Lighthouse                  0        0   
3     Kukailimoku Point Lighthouse                  0        0   
4     Kukailimoku Point Lighthouse                  0        0   
5     Kukailimoku Point Lighthouse                  0        0   
6     Kukailimoku Point Lighthouse                  0        0   
7     Kukailimoku Point Lighthouse                  0        0   
8     Kukailimoku Point Lighthouse                  0        0   
9                 Old Kona Airport                  0        0   
10                Old Kona Airport                  0        0   
11                Old Kona Airport                  0        0   
12                Old Kona Airport                  0        0   
13                Old Kona Airport                  0        0   
14                Old Kona Airport                  0        0   
15  Ellison S Onizuka Space Center                  0        1   
16  Ellison S Onizuka Space Center                  0        0   
17  Ellison S Onizuka Space Center                  0        0   
18  Ellison S Onizuka Space Center                  0        0   
19  Ellison S Onizuka Space Center                  0        0   

    Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                 0                 0                    0             0   
1                 0                 0                    0             0   
2                 0                 0                    0             0   
3                 0                 0                    0             0   
4                 0                 0                    0             0   
5                 0                 0                    0             0   
6                 0                 0                    0             0   
7                 0                 0                    0             0   
8                 0                 0                    0             0   
9                 0                 0                    0             0   
10                0                 0                    0             0   
11                0                 0                    0             0   
12                0                 0                    0             0   
13                0                 0                    0             0   
14                0                 0                    0             0   
15                0                 0                    0             0   
16                1                 0                    0             0   
17                0                 0                    0             0   
18                0                 0                    0             0   
19                1                 0                    0             0   

    Arcade  Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0        0            0           0                    0                 0   
1        0            0           0                    0                 0   
2        0            0           0                    0                 0   
3        0            0           0                    0                 0   
4        0            0           0                    0                 0   
5        0            0           0                    0                 0   
6        0            0           0                    0                 0   
7        0            0           0                    0                 0   
8        0            0           0                    0                 0   
9        0            0           0                    0                 0   
10       0            0           0                    0                 0   
11       0            0           0                    0                 0   
12       0            0 

In [188]:
hawaii_onehot.shape

(5705, 283)

In [191]:
hawaii_grouped = hawaii_onehot.groupby('Neighborhood').mean().reset_index()
hawaii_grouped.head()

Neighborhood  Accessories Store  Airport  Airport Service  \
0        (Pvt) Park                0.0      0.0              0.0   
1                11                0.0      0.0              0.0   
2                12                0.0      0.0              0.0   
3                13                0.0      0.0              0.0   
4  A&B Sugar Museum                0.0      0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  Arcade  Art Gallery  \
0               0.0                 0.05           0.0     0.0          0.0   
1               0.0                 0.00           0.0     0.0          0.0   
2               0.0                 0.00           0.0     0.0          0.0   
3               0.0                 0.00           0.0     0.0          0.0   
4               0.0                 0.00           0.0     0.0          0.0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0         0.0                  0.0               0.0                 0.0   
1         0.0                  0.0               0.0                 0.0   
2         0.0                  0.0               0.0                 0.0   
3         0.0                  0.0               0.0                 0.0   
4         0.0                  0.0               0.0                 0.0   

   Auto Workshop  Automotive Shop  BBQ Joint  Baby Store  Bagel Shop  \
0            0.0             0.05        0.0         0.0         0.0   
1            0.0             0.00        0.0         0.0         0.0   
2            0.0             0.00        0.0         0.0         0.0   
3            0.0             0.00        0.5         0.0         0.0   
4            0.0             0.00        0.0         0.0         0.0   

     Bakery  Bar  Baseball Field  Baseball Stadium  Basketball Court  Bay  \
0  0.000000  0.0             0.0               0.0               0.0  0.0   
1  0.000000  0.0             0.0               0.0               0.0  0.0   
2  0.000000  0.0             0.0               0.0               0.0  0.0   
3  0.000000  0.0             0.0               0.0               0.0  0.0   
4  0.066667  0.0             0.0               0.0               0.0  0.0   

   Beach  Beach Bar  Bed & Breakfast  Beer Bar  Beer Garden  Big Box Store  \
0    0.0        0.0              0.0       0.0          0.0       0.000000   
1    0.0        0.0              1.0       0.0          0.0       0.000000   
2    0.0        0.0              1.0       0.0          0.0       0.000000   
3    0.0        0.0              0.0       0.0          0.0       0.000000   
4    0.0        0.0              0.0       0.0          0.0       0.066667   

   Board Shop  Boat or Ferry  Bookstore  Botanical Garden  Boutique  \
0         0.0            0.0        0.0               0.0       0.0   
1         0.0            0.0        0.0               0.0       0.0   
2         0.0            0.0        0.0               0.0       0.0   
3         0.0            0.0        0.0               0.0       0.0   
4         0.0            0.0        0.0               0.0       0.0   

   Bowling Alley  Breakfast Spot  Brewery  Bridal Shop  Bubble Tea Shop  \
0            0.0             0.0      0.0          0.0              0.0   
1            0.0             0.0      0.0          0.0              0.0   
2            0.0             0.0      0.0          0.0              0.0   
3            0.0             0.0      0.0          0.0              0.0   
4            0.0             0.0      0.0          0.0              0.0   

   Building  Burger Joint  Burrito Place  Bus Station  Bus Stop  \
0       0.0           0.0            0.0          0.1      0.05   
1       0.0           0.0            0.0          0.0      0.00   
2       0.0           0.0            0.0          0.0      0.00   
3       0.0           0.0            0.0          0.0      0.00   
4       0.0           0.0            0.0          0.0      0.00   

   Business Service 

In [192]:
hawaii_grouped.shape

(394, 283)

In [193]:
num_top_venues = 10
for hood in hawaii_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = hawaii_grouped[hawaii_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----(Pvt) Park----
               venue  freq
0     History Museum  0.20
1      Historic Site  0.15
2        Bus Station  0.10
3      Grocery Store  0.05
4        Video Store  0.05
5    Harbor / Marina  0.05
6               Park  0.05
7              Hotel  0.05
8  Convenience Store  0.05
9    Automotive Shop  0.05


----11----
                           venue  freq
0                Bed & Breakfast   1.0
1              Accessories Store   0.0
2                  National Park   0.0
3          Performing Arts Venue   0.0
4                           Park   0.0
5  Paper / Office Supplies Store   0.0
6                   Outlet Store   0.0
7           Outdoor Supply Store   0.0
8              Outdoor Sculpture   0.0
9            Outdoor Event Space   0.0


----12----
                           venue  freq
0                Bed & Breakfast   1.0
1              Accessories Store   0.0
2                  National Park   0.0
3          Performing Arts Venue   0.0
4                           Park  

In [194]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [209]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = hawaii_grouped['Neighborhood']

for ind in np.arange(hawaii_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(hawaii_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0        (Pvt) Park        History Museum         Historic Site   
1                11       Bed & Breakfast                   Zoo   
2                12       Bed & Breakfast                   Zoo   
3                13             BBQ Joint                Hostel   
4  A&B Sugar Museum           Coffee Shop    Salon / Barbershop   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0            Bus Station     Electronics Store                  Park   
1  General Entertainment                  Food     Food & Drink Shop   
2  General Entertainment                  Food     Food & Drink Shop   
3                    Zoo                Forest           Flower Shop   
4                  Diner           Salad Place        Hardware Store   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0         Grocery Store  Fast Food Restaurant       Automotive Shop   
1            Food Court            Food Truck                Forest   
2            Food Court            Food Truck                Forest   
3                  Food     Food & Drink Shop            Food Court   
4                Bakery     Mobile Phone Shop         Big Box Store   

  9th Most Common Venue 10th Most Common Venue  
0     Convenience Store                  Hotel  
1     French Restaurant    Fried Chicken Joint  
2     French Restaurant    Fried Chicken Joint  
3            Food Truck      French Restaurant  
4        History Museum         Sandwich Place

In [211]:
kclusters = 9

hawaii_grouped_clustering = hawaii_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(hawaii_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 4, 4, 0, 2, 2, 2, 1, 2, 1], dtype=int32)

In [224]:
# add clustering labels
# neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

hawaii_merged = neighborhoods

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
hawaii_merged =hawaii_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

hawaii_merged # check the last columns!

Neighborhood   Latitude   Longitude  Cluster Labels  \
0      Kukailimoku Point Lighthouse  19.640560 -156.003466             2.0   
1                  Old Kona Airport  19.647014 -156.011014             6.0   
2    Ellison S Onizuka Space Center  19.738103 -156.043785             2.0   
3          Keahole Point Lighthouse  19.730984 -156.063283             2.0   
4       Honokohau Small Boat Harbor  19.673002 -156.025103             1.0   
5                    Kailua Airport  19.647014 -156.011014             6.0   
6                       Water Tower  19.676486 -156.006658             NaN   
7               Kealakekua Bay Park  19.478702 -155.921985             2.0   
8                 Kalahiki Cemetery  19.378003 -155.878344             NaN   
9                    Hookena School  19.390148 -155.881706             NaN   
10                  Honaunau School  19.453100 -155.881097             2.0   
11                      Alae School  19.332269 -155.876293             NaN   
12   Seamountain Ninole Golf Course  19.137274 -155.511145             2.0   
13           Kaalaiki Landing Strip  19.125878 -155.579529             NaN   
14       Upper Paauau Landing Strip  19.231650 -155.508958             NaN   
15           Kaalaiki Landing Strip  19.124004 -155.581519             NaN   
16                     Naalehu Park  19.063949 -155.585591             2.0   
17                    Waiohinu Park  19.067627 -155.608703             2.0   
18    Whittington Beach County Park  19.088420 -155.553528             6.0   
19                    Mookini Heiau  20.261218 -155.880068             3.0   
20    Kahola Historical Sites State  20.261218 -155.880068             3.0   
21                    Upolu Airport  20.269011 -155.860285             NaN   
22         Kauhola Point Lighthouse  20.249070 -155.773223             NaN   
23                  Kohala Hospital  20.235929 -155.798770             3.0   
24                   Halaula School  20.230600 -155.781744             6.0   
25                 Wainaia Cemetery  20.232458 -155.790555             NaN   
26                    Waimea School  20.022861 -155.672277             2.0   
27                    Parker School  20.025564 -155.672325             2.0   
28               Laupahoehoe School  19.982727 -155.230532             2.0   
29                   Hakalau School  19.891776 -155.127845             NaN   
30                  Waikaumalo Park  19.934818 -155.165765             3.0   
31        Mauna Kea Science Reserve  19.824991 -155.471990             5.0   
32                Pepeekeo Airstrip  19.848971 -155.115091             2.0   
33                     Andrade Camp  19.841755 -155.109342             NaN   
34            Waiakea Middle School  19.699361 -155.079949             8.0   
35                     Wainaku Camp  19.746773 -155.111575             2.0   
36                       Lehia Park  19.737560 -155.010354             6.0   
37               Kealoha Beach Park  19.736654 -155.029962             6.0   
38                      Post Office  19.621999 -155.951204             2.0   
39       Holualoa Elementary School  19.618879 -155.951571             0.0   
40                 Holualoa Library  19.621114 -155.951407             0.0   
41            Konawaena High School  19.513526 -155.917550             2.0   
42                    Kona Hospital  19.522743 -155.921403             2.0   
43                 Higashihara Park  19.554625 -155.939652             2.0   
44          Mauna Lani Point Resort  19.942333 -155.878681             2.0   
45            Mauna Loa Observatory  19.539166 -155.579313             NaN   
46     Kulani Correctional Facility  19.565735 -155.323895             NaN   
47            Mauna Loa Boys School  19.585732 -155.339780             NaN   
48                     Keaau School  19.624588 -155.040633             2.0   
49                 Main Post Office  19.719526 -155.052392             2.0   
50                      Dolphin Bay  19.73197

In [231]:
hawaii_merged.dropna(axis = 0,inplace  = True)
hawaii_merged.head(130)

Neighborhood   Latitude   Longitude  Cluster Labels  \
0      Kukailimoku Point Lighthouse  19.640560 -156.003466             2.0   
1                  Old Kona Airport  19.647014 -156.011014             6.0   
2    Ellison S Onizuka Space Center  19.738103 -156.043785             2.0   
3          Keahole Point Lighthouse  19.730984 -156.063283             2.0   
4       Honokohau Small Boat Harbor  19.673002 -156.025103             1.0   
5                    Kailua Airport  19.647014 -156.011014             6.0   
7               Kealakekua Bay Park  19.478702 -155.921985             2.0   
10                  Honaunau School  19.453100 -155.881097             2.0   
12   Seamountain Ninole Golf Course  19.137274 -155.511145             2.0   
16                     Naalehu Park  19.063949 -155.585591             2.0   
17                    Waiohinu Park  19.067627 -155.608703             2.0   
18    Whittington Beach County Park  19.088420 -155.553528             6.0   
19                    Mookini Heiau  20.261218 -155.880068             3.0   
20    Kahola Historical Sites State  20.261218 -155.880068             3.0   
23                  Kohala Hospital  20.235929 -155.798770             3.0   
24                   Halaula School  20.230600 -155.781744             6.0   
26                    Waimea School  20.022861 -155.672277             2.0   
27                    Parker School  20.025564 -155.672325             2.0   
28               Laupahoehoe School  19.982727 -155.230532             2.0   
30                  Waikaumalo Park  19.934818 -155.165765             3.0   
31        Mauna Kea Science Reserve  19.824991 -155.471990             5.0   
32                Pepeekeo Airstrip  19.848971 -155.115091             2.0   
34            Waiakea Middle School  19.699361 -155.079949             8.0   
35                     Wainaku Camp  19.746773 -155.111575             2.0   
36                       Lehia Park  19.737560 -155.010354             6.0   
37               Kealoha Beach Park  19.736654 -155.029962             6.0   
38                      Post Office  19.621999 -155.951204             2.0   
39       Holualoa Elementary School  19.618879 -155.951571             0.0   
40                 Holualoa Library  19.621114 -155.951407             0.0   
41            Konawaena High School  19.513526 -155.917550             2.0   
42                    Kona Hospital  19.522743 -155.921403             2.0   
43                 Higashihara Park  19.554625 -155.939652             2.0   
44          Mauna Lani Point Resort  19.942333 -155.878681             2.0   
48                     Keaau School  19.624588 -155.040633             2.0   
49                 Main Post Office  19.719526 -155.052392             2.0   
50                      Dolphin Bay  19.731972 -155.093281             2.0   
51              Maui's Canoe (Rock)  19.730174 -155.092140             2.0   
52          Federal Office Building  19.728780 -155.091684             2.0   
53                          Library  19.727733 -155.093594             2.0   
54                 Hilo High School  19.725542 -155.097402             4.0   
55     East Hawai'i Cultural Center  19.727599 -155.090801             2.0   
56          Hawai'i Visitors Bureau  19.726607 -155.089603             2.0   
57              Haili Church (1824)  19.725051 -155.092540             2.0   
58   Lyman Museum and Mission House  19.725239 -155.093822             2.0   
59             Puainako Town Center  19.695160 -155.070059             2.0   
60              Waiakea High School  19.703064 -155.080130             8.0   
61   Hawai'i State Department of Ag  19.709193 -155.076265             2.0   
62   Prince Kuhio Plaza Shop Center  19.700135 -155.067009             2.0   
63            Hawai'i County Police  19.718593 -155.088193             2.0   
64                     Kaiko'o Mall  19.720969 -155.081735             2.0   
65            State Office Building  19.72137

In [235]:
#create map
map_clusters = folium.Map(location=[Latitude, Longitude], zoom_start=10)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(hawaii_merged['Latitude'], hawaii_merged['Longitude'], hawaii_merged['Neighborhood'], hawaii_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters